java-checks/src/main/java/org/sonar/java/checks/ArrayCopyLoopCheck.java

```diff
@@ -0,0 +1,309 @@
+    if (expression.is(Kind.ASSIGNMENT)) {
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) expression;
+      ExpressionTree lhs = assignment.variable();
+      ExpressionTree rhs = assignment.expression();
+      if (lhs.is(Kind.ARRAY_ACCESS_EXPRESSION) && rhs.is(Kind.ARRAY_ACCESS_EXPRESSION)) {
+        ArrayAccessExpressionTree src = (ArrayAccessExpressionTree) rhs;
+        ArrayAccessExpressionTree dst = (ArrayAccessExpressionTree) lhs;
+        return isIdentifier(src.expression()) && useCounter(src, ctx) && isIdentifier(dst.expression()) && useCounter(dst, ctx);
+      }
+    }
+    return false;
+  }
+
+  private static boolean isArrayToListCopy(ExpressionTree expression, Context ctx) {
+    if (expression.is(Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree invocation = (MethodInvocationTree) expression;
+      if (invocation.methodSelect().is(Kind.MEMBER_SELECT) && invocation.arguments().size() == 1) {
+        MemberSelectExpressionTree select = (MemberSelectExpressionTree) invocation.methodSelect();
+        ExpressionTree object = select.expression();
+        IdentifierTree method = select.identifier();
+        if (isIdentifier(object) && isCollection(object) && "add".equals(method.symbol().name())) {
+          ExpressionTree argument = invocation.arguments().get(0);
+          if (argument.is(Kind.ARRAY_ACCESS_EXPRESSION)) {
+            ArrayAccessExpressionTree access = (ArrayAccessExpressionTree) argument;
+            return isIdentifier(access.expression()) && useCounter(access, ctx);
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean isArrayToListCopy(ExpressionTree expression, VariableTree iterated) {
+    if (expression.is(Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree invocation = (MethodInvocationTree) expression;
+      if ("add".equals(invocation.symbol().name()) && invocation.arguments().size() == 1) {
+        expression = invocation.methodSelect();
+        if (expression.is(Kind.MEMBER_SELECT)) {
+          MemberSelectExpressionTree select = (MemberSelectExpressionTree) expression;
+          ExpressionTree object = select.expression();
+          if (isIdentifier(object) && isCollection(object)) {
+            ExpressionTree argument = invocation.arguments().get(0);
+            IdentifierTree identifier = getIdentifier(argument);
+            return identifier != null && identifier.symbol().equals(iterated.symbol());
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean isArray(ExpressionTree tree) {
+    return tree.symbolType().isArray();
+  }
+
+  private static boolean isCollection(ExpressionTree tree) {
+    return tree.symbolType().isSubtypeOf("java.util.Collection");
+  }
+
+  private static boolean isIdentifier(ExpressionTree tree) {
+    return tree.is(Kind.IDENTIFIER);
+  }
+
+  private static IdentifierTree getIdentifier(ExpressionTree tree) {
```

Comment:
When  you use this method, you subsequently always need `Symbol` from the identifier, it would be more usefull to return the `Symbol` directly. Also annotate this method with `@CheckForNull`